src/gpu/locks.rs

```diff
@@ -96,30 +112,32 @@ macro_rules! locked_kernel {
                 }
             }
 
-            pub fn with<F, R>(&mut self, f: F) -> GPUResult<R>
+            pub fn with<F, R>(&mut self, mut f: F) -> GPUResult<R>
             where
-                F: FnOnce(&mut $kern<E>) -> GPUResult<R>,
+                F: FnMut(&mut $kern<E>) -> GPUResult<R>,
             {
-                if PriorityLock::should_break(self.priority) {
-                    self.free();
-                } else if self.kernel.is_none() {
-                    info!("GPU is available for {}!", $name);
-                    self.kernel = $func::<E>(self.log_d, self.priority);
-                }
+                self.init();
 
-                if let Some(ref mut k) = self.kernel {
-                    match f(k) {
-                        Err(e) => {
-                            if let GPUError::GPUTaken = e {
-                                self.free();
+                loop {
+                    if let Some(ref mut k) = self.kernel {
+                        match f(k) {
+                            Err(e) => {
+                                if let GPUError::GPUTaken = e {
```

Comment:
Before even trying this out. I think you could move this `if let` directly into the matching clause, something like:

```rust
match f(k) {
    Err(GPUError::GPUTaken) => …
    Err(error) => …
    Ok(v) => …
}

And another note. I personally find `if let … else` clauses strange. Most of the time I find using matching clauses in such cases instead. I use `if let` only if there is no `else` clause.

Reply:
thanks, it looks much nicer now